distributed/shuffle/multi_comm.py

```diff
@@ -0,0 +1,215 @@
+import asyncio
+import contextlib
+import threading
+import time
+import weakref
+from collections import defaultdict
+
+from dask.utils import parse_bytes
+
+from distributed.utils import log_errors
+
+
+class MultiComm:
+    """Accept, buffer, and send many small messages to many workers
+
+    This takes in lots of small messages destined for remote workers, buffers
+    those messages in memory, and then sends out batches of them when possible
+    to different workers.  This tries to send larger messages when possible,
+    while also respecting a memory bound
+
+    **State**
+
+    -   shards: dict[str, list[bytes]]
+
+        This is our in-memory buffer of data waiting to be sent to other workers.
+
+    -   sizes: dict[str, int]
+
+        The size of each list of shards.  We find the largest and send data from that buffer
+
+    State
+    -----
+
+    memory_limit: str
+        A maximum amount of memory to use, like "1 GiB"
+    max_connections: int
+        The maximum number of connections to have out at once
+    max_message_size: str
+        The maximum size of a single message that we want to send
+
+    Parameters
+    ----------
+    send: callable
+        How to send a list of shards to a worker
+    """
+
+    max_message_size = parse_bytes("2 MiB")
+    memory_limit = parse_bytes("100 MiB")
+    max_connections = 10
+    _queues: weakref.WeakKeyDictionary = weakref.WeakKeyDictionary()
+    total_size = 0
+    lock = threading.Lock()
+
+    def __init__(
+        self,
+        send=None,
+        loop=None,
+    ):
+        self.send = send
+        self.shards = defaultdict(list)
+        self.sizes = defaultdict(int)
+        self.total_size = 0
+        self.total_moved = 0
+        self.thread_condition = threading.Condition()
+        self._futures = set()
+        self._done = False
+        self.diagnostics = defaultdict(float)
+        self._loop = loop or asyncio.get_event_loop()
+
+        self._communicate_future = asyncio.create_task(self.communicate())
+        self._exception = None
+
+    @property
+    def queue(self):
+        try:
+            return MultiComm._queues[self._loop]
+        except KeyError:
+            queue = asyncio.Queue()
+            for _ in range(MultiComm.max_connections):
+                queue.put_nowait(None)
```

Comment:
It took me a bit to understand that this queue works the opposite way you'd expect—it's not pushing data in, it's pushing permission to write more back out. Sort of a form of [`CapacityLimiter`](https://trio.readthedocs.io/en/stable/reference-core.html#trio.CapacityLimiter)? I think this would help to document.

Reply:
I agree with the comment. Took me a bit to understand this myself. 

I think factoring this out should not be scope of this PR but I could see value in doing it regardless.

I think this is a case for a follow up ticket. @gjoseph92 would you mind opening one?